test/parallel_api/numeric/numeric.ops/exclusive_scan_by_segment.pass.cpp

```diff
@@ -20,146 +20,194 @@
     {
-        auto key_buf = _key_buf.template get_access<sycl::access::mode::read_write>();
-        auto val_buf = _val_buf.template get_access<sycl::access::mode::read_write>();
-        auto res_buf = _res_buf.template get_access<sycl::access::mode::read_write>();
-
-        // Initialize data
-        key_buf[0] = 0; key_buf[1] = 0; key_buf[2] = 0; key_buf[3] = 1; key_buf[4] = 1;
-        key_buf[5] = 2; key_buf[6] = 3; key_buf[7] = 3; key_buf[8] = 3; key_buf[9] = 3;
-
-        val_buf[0] = 1; val_buf[1] = 1; val_buf[2] = 1; val_buf[3] = 1; val_buf[4] = 1;
-        val_buf[5] = 1; val_buf[6] = 1; val_buf[7] = 1; val_buf[8] = 1; val_buf[9] = 1;
-
-        res_buf[0] = 9; res_buf[1] = 9; res_buf[2] = 9; res_buf[3] = 9; res_buf[4] = 9;
-        res_buf[5] = 9; res_buf[6] = 9; res_buf[7] = 9; res_buf[8] = 9; res_buf[9] = 9;
+        //T keys[n1] = { 1, 2, 3, 4, 1, 1, 2, 2, 3, 3, 4, 4, 1, 1, 1, ...};
+        //T vals[n1] = { 1, 1, 1, ... };
+        const typename std::decay<decltype(host_vals[0])>::type value = 1;
+
+        int segment_length = 1;
+        int i = 0;
+        while (i != n)
+        {
+          for (int j = 0; j != 4*segment_length && i != n; ++j)
+          {
+              host_keys[i] = j/segment_length + 1;
+              host_vals[i] = 1;
+              host_val_res[i] = 0;
+              ++i;
+          }
+          ++segment_length;
+        }
     }
 
-    // create sycl iterators
-    auto key_beg = oneapi::dpl::begin(_key_buf);
-    auto key_end = oneapi::dpl::end(_key_buf);
-    auto val_beg = oneapi::dpl::begin(_val_buf);
-    auto res_beg = oneapi::dpl::begin(_res_buf);
-
-    // create named policy from existing one
-    auto new_policy = oneapi::dpl::execution::make_device_policy<class ExclusiveScanBySegment>(oneapi::dpl::execution::dpcpp_default);
-    // call algorithm
-    oneapi::dpl::exclusive_scan_by_segment(new_policy, key_beg, key_end, val_beg, res_beg,
-        (uint64_t)0, std::equal_to<uint64_t>(), std::plus<uint64_t>());
-
-    // check values
-    auto key_acc = _key_buf.get_access<sycl::access::mode::read>();
-    auto val_acc = _val_buf.get_access<sycl::access::mode::read>();
-    auto res_acc = _res_buf.get_access<sycl::access::mode::read>();
-    uint64_t check_value;
-    for (int i = 0; i != 10; ++i) {
-            if (i == 0 || key_acc[i] != key_acc[i-1])
-                check_value = 0;
-        else
-        check_value += val_acc[i-1];
-        ASSERT_EQUAL(check_value, res_acc[i]);
+    template <typename Accessor1, typename Accessor2, typename T, typename Size>
+    void
+    check_values(Accessor1 host_keys, Accessor2 val_res, T init, Size n)
+    {
+        //T keys[n1] = { 1, 2, 3, 4, 1, 1, 2, 2, 3, 3, 4, 4, 1, 1, 1, ...};
+        //T vals[n1] = { 1, 1, 1, 1, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 3, ...};
+
+        int segment_length = 1;
+        auto expected_segment_sum = init == 1 ? segment_length * (segment_length + 1) / 2 : segment_length * (segment_length - 1) / 2;
```

Comment:
Can we use ```auto expected_segment_sum = init``` instead? It seems the check will work correctly in case init is 0 or 1. Let's add at least an assert that we get expected init.

Reply:
Yes, using `init` works.  I've added the assert as well.